samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -56,53 +54,75 @@
-  private static final String LATCH_INIT = "init";
-  private static final long LATCH_TIMEOUT_MINUTES = 10; // 10 min timeout
+  // Latch id that's used for awaiting the init of application before creating the StreamProcessors
+  private static final String INIT_LATCH_ID = "init";
+  // Latch timeout is set to 10 min
+  private static final long LATCH_TIMEOUT_MINUTES = 10;
 
   private final String uid;
-  private final CoordinationUtils coordination;
-  private final List<StreamProcessor> processors = new ArrayList<>();
-  private final CountDownLatch latch = new CountDownLatch(1);
-  private final AtomicReference<Throwable> throwable = new AtomicReference<>();
-  private final ConcurrentHashSet<String> processorIds = new ConcurrentHashSet<>();
+  private final CoordinationUtils coordinationUtils;
+  private final Set<StreamProcessor> processors = ConcurrentHashMap.newKeySet();
+  private final CountDownLatch shutdownLatch = new CountDownLatch(1);
+  private final AtomicInteger numProcessorsToStart = new AtomicInteger();
 
   private ApplicationStatus appStatus = ApplicationStatus.New;
 
-  final class LocalStreamProcessorListener implements StreamProcessorLifecycleListener {
-    public final String processorId;
+  final class LocalStreamProcessorLifeCycleListener implements StreamProcessorLifecycleListener {
+    StreamProcessor processor;
 
-    public LocalStreamProcessorListener(String processorId) {
-      if (StringUtils.isEmpty(processorId)) {
-        throw new NullPointerException("processorId has to be defined in LocalStreamProcessorListener.");
-      }
-      this.processorId = processorId;
+    void setProcessor(StreamProcessor processor) {
+      this.processor = processor;
     }
 
     @Override
     public void onStart() {
-      processorIds.add(processorId);
+      if (numProcessorsToStart.decrementAndGet() == 0) {
+        appStatus = ApplicationStatus.Running;
+      }
     }
 
     @Override
     public void onShutdown() {
-      processorIds.remove(processorId);
-      if (processorIds.isEmpty()) {
-        latch.countDown();
+      processors.remove(processor);
+      processor = null;
+
+      if (processors.isEmpty()) {
+        if (appStatus == ApplicationStatus.Running) {
+          appStatus = ApplicationStatus.SuccessfulFinish;
+        } else if (appStatus == ApplicationStatus.New) {
+          // the processor is shutdown before started
+          appStatus = ApplicationStatus.UnsuccessfulFinish;
+        }
+        shutdownAndNotify();
       }
     }
 
     @Override
     public void onFailure(Throwable t) {
-      processorIds.remove(processorId);
-      throwable.compareAndSet(null, t);
-      latch.countDown();
+      processors.remove(processor);
+      processor = null;
+
+      appStatus = ApplicationStatus.unsuccessfulFinish(t);
+      if (!processors.isEmpty()) {
+        // shut down the other processors
+        processors.forEach(StreamProcessor::stop);
```

Comment:
I think it should be fine. It might invoke jc.stop twice. But I think the current implementations of JC don't cause any state change or failure on invoking stop multiple times.. So we should be fine, **I think**